### ISSUE TITLE
Messes up '.

### ISSUE THREAD

**Issue by krzkrzkrz** [rust-lang/rust.vim#287]
Every time a save is done. It messages up `'.`

As in. In Vim. When I do `'.`, Vim always takes the cursor to the end of the file, instead of the last changed mark

I think it has something to do with `rustfmt`

Is this the expected behaviour or is there an option I need to set in `.vimrc`?

**Comment by da-x** [comment]
Preserving marks is indeed an issue when reformatting replaces the entire buffer content. Perhaps, if not much as changed during the reformatting, we can diff the before-and-after versions and adjust the mark accordingly. I don't see handling for this in other formatters like [neoformat](https://github.com/sbdchd/neoformat), though.

**Comment by krzkrzkrz** [comment]
Remembering/retaining marks would be a good addition

**Comment by da-x** [comment]
I agree, though adjusting the marks on the correct line post-formatting would not be exact without knowing precisely how tokens have shifted during the reformatting. Only rustfmt retains that exact information.

Checking with @nrc - perhaps rustfmt can be given an optional list of `(line, column)` positions, and then it would emit the new positions after formatting? Something like, `--marks-in [filename]`, `--marks-out [filename]`, where file format would be a YAML serialization of `[(String, u64, u64)]`. But perhaps we can come up with a better protocol.

In the future, this will also be helpful in adjusting not only marks, but buffer highlights too.

**Comment by da-x** [comment]
Move to end of file is fixed now by #419 perhaps, want to check?

**Comment by benknoble** [comment]
Could we use `:keepmarks` on the vim sides? It currently only has effect for `:[range]!`, according to the docs.
